### ISSUE TITLE
Plain identifier does not work for an entity

### ISSUE THREAD

**Issue by Soullivaneuh** [api-platform/core#2320]
I have enabled the `allow_plain_identifiers`, and it works fine on `POST` for most of my entities... except this one so far (I reduced to the relation only):

```php
/**
 * @ApiResource(
 *     shortName="DnsRecord",
 *     collectionOperations={"post"}
 * )
 * @OwnerAware(fields={"domain"})
 */
class PowerDNSRecord implements OwnableInterface
{
    /**
     * @var PowerDNSDomain
     *
     * @ORM\ManyToOne(targetEntity="PowerDNSBundle\Entity\PowerDNSDomain", inversedBy="records", fetch="EAGER")
     * @ORM\JoinColumn(name="domain_id", referencedColumnName="id", onDelete="CASCADE")
     * @Groups("post")
     */
    private $domain;

    public function isOwnedBy(User $user): bool
    {
        return $this->domain->isOwnedBy($user);
    }

    public function setDomain(PowerDNSDomain $domain): self
    {
        $this->domain = $domain;

        return $this;
    }

    public function getDomain(): PowerDNSDomain
    {
        return $this->domain;
    }
}
```

This entity belongs to this one:

```php
/**
 * @ApiResource(
 *     shortName="DnsZone",
 *     collectionOperations={"get", "post"},
 *     itemOperations={"get", "delete"}
 * )
 * @OwnerAware
 */
class PowerDNSDomain implements OwnableInterface
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer", nullable=false)
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="IDENTITY")
     *
     * @Groups({"gets", "delete"})
     */
    private $id;

    /**
     * @var string
     *
     * @ORM\Column(name="name", type="string", length=255, nullable=false)
     * @Assert\NotBlank
     * @Assert\Regex("/^[[:alnum:]-\.]+$/u")
     *
     * @Groups({"gets", "post"})
     */
    private $name;

    /**
     * @var int|null
     *
     * @ORM\Column(name="owner_id", type="integer", nullable=true)
     *
     * @Groups("gets:admin")
     */
    private $ownerId;

    /**
     * @var PowerDNSRecord[]|ArrayCollection
     *
     * @ORM\OneToMany(targetEntity="PowerDNSRecord", mappedBy="domain")
     * @ApiSubresource
     */
    private $records;

    public function __construct()
    {
        $this->records = new ArrayCollection();
    }

    public function isOwnedBy(User $user): bool
    {
        return $user->getId() === $this->ownerId;
    }

    public function setName(string $name): self
    {
        $this->name = $name;

        return $this;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function addRecord(PowerDNSRecord $records): self
    {
        $this->records[] = $records;

        return $this;
    }

    public function removeRecord(PowerDNSRecord $records): void
    {
        $this->records->removeElement($records);
    }

    /**
     * @return PowerDNSRecord[]|ArrayCollection
     */
    public function getRecords(): Collection
    {
        return $this->records;
    }
}
```

If I use plain identifier, api-platform tries to create a new `PowerDNSDomain` entities. It works fine if I use the IRI format.

I suspected the `shortName` first, but it does not change a thing if I remove it.

I don't have anymore clue. Please tell me if you need more information.

**Comment by dunglas** [comment]
The current implementation of plain identifiers is buggy and not reliable. I would like to deprecate it and replace it by what I've described here: https://github.com/api-platform/core/pull/2022#issuecomment-398851861

However I lack of time to work on this :(. Any help would be very welcome.

**Comment by Soullivaneuh** [comment]
Right, I will read all the comments next time: https://github.com/api-platform/core/pull/2022#issuecomment-398861586

So if I well understand I have to look to the ItemNormalizer to make some debug?

Not sure if I'll be able to make what you want in https://github.com/api-platform/core/pull/2022#issuecomment-398851861, but I still have to look for a fix. :wink:

**Comment by Soullivaneuh** [comment]
Ok so I found the difference between the working one and the not working one.

It's the `PropertyMetadata::isWritableLink`.

Indeed, for the `PowerDNSRecord` creation, `isWritableLink` is set to true due to serialization group intersect.

So it try to create a new entity here: https://github.com/api-platform/core/blob/516c9c701ec5bfe15b7587a3edda95b2dd9cbb30/src/Serializer/AbstractItemNormalizer.php#L372-L389

We can see that it let the code continue if the creation fail but that's the issue: The entity creation "works" here.

It should go here instead: https://github.com/api-platform/core/blob/516c9c701ec5bfe15b7587a3edda95b2dd9cbb30/src/Serializer/AbstractItemNormalizer.php#L391-L406

Two possible solutions:

1. Reverse the condition order, try to get the relation first and then create it if possible.
2. Add a condition constraint on the creation part to do nothing is the value is not an array.

What do you think? What possible side effect may I have?

**Comment by Soullivaneuh** [comment]
Works with this:

```php
if (
    \is_array($value)
    && (
        !$this->resourceClassResolver->isResourceClass($className) ||
        $propertyMetadata->isWritableLink()
    )
) {
```

And looks legit to me, we should not create a relation based only on a string or integer value, WDYT?

**Comment by Soullivaneuh** [comment]
Closes for #2200.
